src/ray/gcs/asio.cc

```diff
@@ -65,62 +65,49 @@ void RedisAsioClient::operate() {
   if (read_requested_ && !read_in_progress_) {
     read_in_progress_ = true;
     socket_.async_read_some(boost::asio::null_buffers(),
-                            boost::bind(&RedisAsioClient::handle_read, this,
-                                        boost::asio::placeholders::error));
+                            boost::bind(&RedisAsioClient::handle_io, this,
+                                        boost::asio::placeholders::error, false));
   }
 
   if (write_requested_ && !write_in_progress_) {
     write_in_progress_ = true;
     socket_.async_write_some(boost::asio::null_buffers(),
-                             boost::bind(&RedisAsioClient::handle_write, this,
-                                         boost::asio::placeholders::error));
+                             boost::bind(&RedisAsioClient::handle_io, this,
+                                         boost::asio::placeholders::error, true));
   }
 }
 
-void RedisAsioClient::handle_read(boost::system::error_code error_code) {
+void RedisAsioClient::handle_io(boost::system::error_code error_code, bool write) {
   RAY_CHECK(!error_code || error_code == boost::asio::error::would_block ||
-            error_code == boost::asio::error::connection_reset)
-      << "handle_read(error_code = " << error_code << ")";
-  read_in_progress_ = false;
-  redis_async_context_.RedisAsyncHandleRead();
-
-  if (error_code == boost::asio::error::would_block) {
-    operate();
+            error_code == boost::asio::error::connection_reset ||
+            error_code == boost::asio::error::operation_aborted)
+      << "handle_io(error_code = " << error_code << ")";
+  (write ? write_in_progress_ : read_in_progress_) = false;
+  if (error_code != boost::asio::error::operation_aborted) {
+    if (!redis_async_context_.GetRawRedisAsyncContext()) {
```

Comment:
I had `RAY_CHECK` in the beginning, but it hindered my ability to set a breakpoint for when the condition is false. So it'd be nice to use `RAY_LOG(FATAL)` for now, since this code might still need further debugging before all issues are resolved (I think it might still be buggy).